### PR TITLE
[libc][math][c23] adds `nanf128`

### PR DESCRIPTION
**Pull request by ashfordium**
Continuing  #84689, this one required more changes than the others, so I am making it a separate PR.

Extends some stuff in `str_to_float.h`, `str_to_integer.h` to work on types wider than `unsigned long long` and `uint64_t`.

~~This also edits `UInt.h` to add type traits for `make_unsigned` and `make_signed` since `make_unsigned<UInt<128>>` is used in `strtointeger` now. I also added the other cases of `make_unsigned` and `make_signed` for `BigInt` for completeness.~~ See review comments below.

I think the float128, and other code making use of `UInt.h`, are only being tested with `LIBC_TYPES_HAS_INT128` defined. When I removed the define of `LIBC_TYPES_HAS_INT128` in `src/__support/macros/properties/types.h`, I got a lot of compiler errors. I think this is because many type traits are not defined for BigInts, only for `__uint128_t` and `__int128_t`. `is_unsigned` is one type trait that is not defined for BigInts. I am also unsure if this is an issue or if I am just not 'not defining' `LIBC_TYPES_HAS_INT128` properly. If this is an issue, I also wonder if we should be testing both sides of the preprocessor condition in `UInt128.h`. Let me know if I should open an issue.

cc @lntue for review.

### CODE REVIEW

File: libc/src/__support/UInt.h [hunk 1]
```diff
@@ -993,6 +993,15 @@ struct is_big_int<BigInt<Bits, Signed, T>> : cpp::true_type {};
...
 
+// make_unsigned and make_signed type traits
+template <size_t Bits, bool Signed, typename T>
+struct cpp::make_unsigned<BigInt<Bits, Signed, T>>
+    : cpp::type_identity<BigInt<Bits, false, T>> {};
+
+template <size_t Bits, bool Signed, typename T>
+struct cpp::make_signed<BigInt<Bits, Signed, T>>
+    : cpp::type_identity<BigInt<Bits, true, T>> {};
+
```
Comment (lntue) [inline libc/src/__support/UInt.h:1004]: We were getting away from specializing `make_signed` and `make_unsigned` for `BigInt` class.  For the context, see https://github.com/llvm/llvm-project/pull/84299

For getting signed and unsigned types of `BigInt`, you might refactor https://github.com/llvm/llvm-project/blob/main/libc/src/__support/integer_to_string.h#L153 to add something like: `is_integral_or_bigint`, `make_integral_or_bigint_signed`, `make_integral_or_bigint_unsigned`, together with their `_v` and `_t` templates.

File: libc/config/linux/aarch64/entrypoints.txt [hunk 2]
```diff
@@ -468,6 +468,7 @@ if(LIBC_TYPES_HAS_FLOAT128)
     libc.src.math.lrintf128
     libc.src.math.lroundf128
     libc.src.math.modff128
+    libc.src.math.nanf128
```
Comment (lntue) [inline libc/config/linux/aarch64/entrypoints.txt:471]: Also update status docs.

File: libc/src/__support/str_to_integer.h [hunk 3]
```diff
@@ -101,15 +106,16 @@ strtointeger(const char *__restrict src, int base,
...
-          : cpp::numeric_limits<T>::max();
-  unsigned long long const abs_max =
+
+  ResultType constexpr NEGATIVE_MAX =
+      !IS_UNSIGNED ? static_cast<ResultType>(cpp::numeric_limits<T>::max()) + 1
+                   : cpp::numeric_limits<T>::max();
+  ResultType const abs_max =
       (is_positive ? cpp::numeric_limits<T>::max() : NEGATIVE_MAX);
-  unsigned long long const abs_max_div_by_base = abs_max / base;
+  ResultType const abs_max_div_by_base = abs_max / base;
```
Comment (lntue) [inline libc/src/__support/str_to_integer.h:117]: Make these 2 `constexpr`.

File: libc/test/src/math/smoke/nanf128_test.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+#include "test/UnitTest/FPMatcher.h"
+#include "test/UnitTest/Test.h"
+
+class LlvmLibcNanf128Test : public LIBC_NAMESPACE::testing::Test {
+public:
+  using FPBits128 = LIBC_NAMESPACE::fputil::FPBits<float128>;
+  using StorageType = FPBits128::StorageType;
+
+  const UInt128 QUIET_NAN =
+      FPBits128::quiet_nan(LIBC_NAMESPACE::fputil::Sign::POS, 0).uintval();
```
Comment (lntue) [inline libc/test/src/math/smoke/nanf128_test.cpp:21]: This is using all the default arguments, so you can shorten it to `FPBits128::quiet_nan().uintval()`

File: libc/src/__support/UInt.h (hunk 1 above)
Comment (gchatelet) [reply libc/src/__support/UInt.h]: It's a bit verbose but at least it's pretty clear what it does. LGTM.

**Review by gchatelet — APPROVED** [review]

**Review by lntue — APPROVED** [review]

**Review by lntue — APPROVED** [review]

**Pull request merged**